### PR TITLE
Add multi-gpu and dataloader support for dfencoder

### PR DESCRIPTION
**Pull request by hsin-c**
## Description
This PR makes the necessary updates to dfencoder to support distributed training.
Key changes:
+ Extracts the core pytorch autoencoder module from the AutoEncoder class and puts it into a new class in a new file `ae_module.py`.
+ Adds a dataloader.py file with a custom Dataset class and a custom DataLoader class to support loading data in batches (assuming the data is saved in one folder with many small csv files)
+ Adds functionalities in AutoEncoder to support distributed training
    - Allows initialization of features from the provided feature information instead of requiring the full dataset to be loaded as a pandas dataframe
    - Adds data preprocessing functions to support the usage of a dataloader
    - Implements distributed model building/training/validation/early stopping

### CODE REVIEW

**Comment by efajardo-nv** [comment]
DFP inference stage calls `get_results` in the model pulled from mlflow. This method however is in `AutoEncoder`. The model saved in mlflow is `AEModule` so we get this error:
```
File "/my_data/gitrepos/hsin-c/Morpheus/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_inference_stage.py", line 86, in on_data
    results_df = loaded_model.get_results(df_user, return_abs=True)
  File "/opt/conda/envs/morpheus/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1269, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'AEModule' object has no attribute 'get_results'
```

File: morpheus/models/dfencoder/autoencoder.py [hunk 1]
```diff
@@ -787,6 +810,396 @@ def fit(self, df, epochs=1, val=None, run_validation=False, use_val_for_loss_sta
             i_loss = cce_loss[:, i]
             self.feature_loss_stats[ft] = self._create_stat_dict(i_loss)
 
+    def fit_distributed(
+        self, 
+        train_dataloader, 
+        rank, 
+        world_size,
```
Comment (efajardo-nv) [inline morpheus/models/dfencoder/autoencoder.py:817]: not sure how to set `rank` or `world_size`. can they have default values or more doc to explain how to set them?

Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py:817]: rank and world_size are variables provided to use when we set up multi processes. Sharing a script I used to run this offline.

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
There are a few things that need to be improved with this PR:

1. There are no tests that exercise the new functionality or show how it should be used.
   1. Even machines with a single GPU should be able to run in distributed mode with 1 worker
   1. This also is a chance to show how this code should be used.
1. All of the added docstrings are in the wrong format.
   1. I commented on a few but all need to be updated into NumPyDoc format. See docstrings on other classes. You can use [this](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) extension to auto-populate the docstrings for you
1. We should be upgrading to use the `logger` module instead of `print()` statements
1. I have a few issues with the design of the class. It seems like it would be too easy to accidentally run the class in both single-GPU and multi-GPU mode.
   1. It might be better to use arguments set in the constructor to determine if this is multi-GPU or not. That way it doesnt matter the order you call functions, you will always get the same result.
   1. We can discuss this more offline
1. We should make any function that should only be called by the class private (leading `_`). 
   1. Too many of these class functions should only be called by the class and its confusing. 
   1. For example, who is supposed to call `build_model`? The class, the user of the class, or both?

File: morpheus/models/dfencoder/ae_module.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+
+import torch
+
+
+def compute_embedding_size(n_categories):
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/ae_module.py]: Renamed.

File: morpheus/models/dfencoder/ae_module.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    n_categories is the number of unique categories in a column.
+    """
+    val = min(600, round(1.6 * n_categories**0.56))
+    return int(val)
+
+
+class CompleteLayer(torch.nn.Module):
+    """
+    Impliments a layer with linear transformation
+    and optional activation and dropout."""
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/ae_module.py]: All docstrings in this class updated.

File: morpheus/models/dfencoder/ae_module.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+    def __init__(self, in_dim, out_dim, activation=None, dropout=None, *args, **kwargs):
+        super(CompleteLayer, self).__init__(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/ae_module.py]: Updated.

File: morpheus/models/dfencoder/ae_module.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            raise Exception(msg)
+
+    def forward(self, x):
+        for layer in self.layers:
+            x = layer(x)
+        return x
+
+
+class AEModule(torch.nn.Module): 
+    """ Auto Encoder Pytorch Module."""
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/ae_module.py:120]: All docstrings in this class updated.

File: morpheus/models/dfencoder/ae_module.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        self.encoder = []
+        self.decoder = []
+        self.numeric_output = None
+        self.binary_output = None
+        # mapping cat features to the output layer
+        self.categorical_output = OrderedDict()
+
+    def build(self, numeric_fts, binary_fts, categorical_fts):
+        if self.verbose:
+            print("Building model...")
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/ae_module.py]: Updated to use `logging.info` here.

File: morpheus/models/dfencoder/autoencoder.py [hunk 7]
```diff
@@ -75,70 +78,6 @@ def ohe(input_vector, dim, device="cpu"):
...
-
-    def forward(self, x):
-        for layer in self.layers:
-            x = layer(x)
-        return x
-
-
 class AutoEncoder(torch.nn.Module):
 
     def __init__(
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py:107]: Addressed.

File: morpheus/models/dfencoder/autoencoder.py [hunk 8]
```diff
@@ -253,22 +201,61 @@ def get_scaler(self, name):
...
+        if df is None and self.preset_numerical_scaler_params is None:
+            raise ValueError(
+                "Either `df` or `self.preset_numerical_scaler_params` needs to be provided."
+            )
+
+        if self.preset_numerical_scaler_params:
+            if self.verbose:
+                print(
+                    "Using self.preset_numerical_scaler_params to override the numerical scalers..."
+                )
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py]: All the print statements updated to be logger.debug from the logging lib.

File: morpheus/models/dfencoder/autoencoder.py [hunk 9]
```diff
@@ -253,22 +201,61 @@ def get_scaler(self, name):
...
+                    "std": scaler_params["std"],
+                    "scaler": scaler,
+                }
+                self.numeric_fts[ft] = feature
+        else:
+            # initialize using a dataframe
+            dt = df.dtypes
+            numeric = []
+            numeric += list(dt[dt == int].index)
+            numeric += list(dt[dt == float].index)
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py]: Updated.

File: morpheus/models/dfencoder/autoencoder.py [hunk 10]
```diff
@@ -416,83 +396,73 @@ def prepare_df(self, df):
...
-        Builds autoencoder model.
+        Builds the autoencoder model using either the given dataframe or the preset feature information for metadata.
+        If distributed training is enabled, wraps the pytorch module by DDP.
 
-        Returns the dataframe after making changes.
+        Args:
+            df (dataframe, optional): the input dataframe to be used to infer metadata
+            distributed_training (bool, optional): whether to enable distributed training
+            rank (int, optional): rank of the process being used for distributed training, 
+                used only if distributed_training=True
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py]: All docstrings updated.

File: morpheus/models/dfencoder/autoencoder.py [hunk 11]
```diff
@@ -416,83 +396,73 @@ def prepare_df(self, df):
...
-                params,
-                lr,
-                momentum=self.momentum,
-                nesterov=self.nesterov,
-                dampening=self.dampening,
-                weight_decay=self.weight_decay,
-            )
-
-    def build_model(self, df):
+    def build_model(self, df=None, distributed_training=False, rank=None):
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py]: `distributed_training` now internalized into a class attribute. Please review the update and let me know if there is still a change required.

File: morpheus/models/dfencoder/autoencoder.py [hunk 12]
```diff
@@ -787,6 +810,396 @@ def fit(self, df, epochs=1, val=None, run_validation=False, use_val_for_loss_sta
...
+            )
+
+        if use_val_for_loss_stats and val_dataset is None:
+            raise ValueError(
+                "Validation set is required if either run_validation or \
+                use_val_for_loss_stats is set to True."
+            )
+
+        if self.optim is None:
+            self.build_model(rank=rank)
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py]: Good catch! It should have. 
The parameters of build_model changed as distributed_training is now a class attribute.  This bug should be resolved now.

File: morpheus/models/dfencoder/autoencoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py]: Discussed offline. Added one shared `fit` function as the entry point for both centralized and distributed training. 
Added type check of `train_data` for the `fit` function so we address the scenarios properly:
- Centralized, train_data is a dataframe -- OK
- Centralized, train_data is a dataloader/dataset -- Error
- Distributed, train_data is a dataframe -- OK, convert it to a dataloader that loads from a df
- - Distributed, train_data is a dataloader/dataset -- OK

[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/autoencoder.py]: Will have a follow-up PR to merge the two functions more elegantly.

File: morpheus/models/dfencoder/dataloader.py [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+                for fn in self.filenames
+            }
+
+        self.file_sizes = {
+            fn: self._get_file_len(fn) - 1
+            if not self.preloaded_data
+            else len(self.preloaded_data[fn])
+            for fn in self.filenames
+        }
+        self.len = sum(v for v in self.file_sizes.values())
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/dataloader.py]: Updated.

File: morpheus/models/dfencoder/dataloader.py [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+    )
+    return dataloader
+    
+def get_distributed_training_dataloader_from_dataset(
+    dataset, rank, world_size, pin_memory=False, num_workers=0
+):
+    sampler = DistributedSampler(
+        dataset, num_replicas=world_size, rank=rank, shuffle=True, drop_last=False
+    )
+    dataloader = DFEncoderDataLoader(
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply morpheus/models/dfencoder/dataloader.py]: All helper functions moved into the corresponding classes.

**Comment by hsin-c** [comment]
> There are a few things that need to be improved with this PR:
> 
> 1. There are no tests that exercise the new functionality or show how it should be used.
>    
>    1. Even machines with a single GPU should be able to run in distributed mode with 1 worker
>    2. This also is a chance to show how this code should be used.
> 2. All of the added docstrings are in the wrong format.
>    
>    1. I commented on a few but all need to be updated into NumPyDoc format. See docstrings on other classes. You can use [this](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) extension to auto-populate the docstrings for you
> 3. We should be upgrading to use the `logger` module instead of `print()` statements
> 4. I have a few issues with the design of the class. It seems like it would be too easy to accidentally run the class in both single-GPU and multi-GPU mode.
>    
>    1. It might be better to use arguments set in the constructor to determine if this is multi-GPU or not. That way it doesnt matter the order you call functions, you will always get the same result.
>    2. We can discuss this more offline
> 5. We should make any function that should only be called by the class private (leading `_`).
>    
>    1. Too many of these class functions should only be called by the class and its confusing.
>    2. For example, who is supposed to call `build_model`? The class, the user of the class, or both?

Everything mentioned should be addressed now.
1. Test added.
2. All newly added docstrings are converted to the numpy doc format.
3. autoencoder.py updated to use logger.debug (if self.verbose set log level to debug) instead of print
4. `fit` function updated according to the discussion offline. Will have a follow-up PR to merge the two paths.
5. Updated the functions in the new classes and many of the functions in the `AutoEncoder` class. Will do a more thorough pass on the `AutoEncoder` path in the follow-up PR.

**Comment by efajardo-nv** [comment]
`val` was renamed to `val_data` in `fit` function so this line in `dfp_training.py`:
https://github.com/hsin-c/Morpheus/blob/hsin-multi-gpu-dfencoder/examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_training.py#L96

needs to be updated to this:
```
model.fit(train_df, epochs=self._epochs, val_data=validation_df, run_validation=run_validation)
```

**Review by mdemoret-nv — APPROVED** [review]
Now that CI is passing, things LGTM

**Review by dagardner-nv — APPROVED** [review]

**Review by drobison00 — APPROVED** [review]
Looks like there is one unresolved comment; good otherwise.

**Pull request merged**